File: LeetCode/427.Construct_Quad_Tree/Mario/427.Construct_Quad_Tree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
};
*/

class Solution {
    public Node construct(int[][] grid) {
        return createquad(grid, 0, grid.length - 1, 0, grid.length - 1);
    }

    private Node createquad(int[][] grid, int rs, int re, int cs, int ce) {
        if (rs > re || cs > ce) {
            return null;
        }

        boolean isLeaf = true;
        int val = grid[rs][cs];
        for (int i = rs; i <= re; i++) {
            for (int j = cs; j <= ce; j++) {
                if (grid[i][j] != val) {
                    isLeaf = false;
                    break;
                }
            }
            if (!isLeaf)
                break;
        }

        if (isLeaf)
            return new Node(val == 1, true);

        int rm = (rs + re) / 2;
        int cm = (cs + ce) / 2;
        Node topLeft = createquad(grid, rs, rm, cs, cm);
        Node topRight = createquad(grid, rs, rm, cm + 1, ce);
        Node bottomLeft = createquad(grid, rm + 1, re, cs, cm);
        Node bottomRight = createquad(grid, rm + 1, re, cm + 1, ce);

        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }
}

File: LeetCode/427.Construct_Quad_Tree/Mario/427.Construct_Quad_Tree.py
#!/usr/bin/env python3

"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""


class Solution:
    def construct(self, grid: List[List[int]]) -> "Node":
        # divide and conquer
        # x: row, y: col
        def build(xi, yi, xj, yj):
            if xi > xj or yi > yj:
                return None

            is_leaf = True
            val = grid[xi][yj]
            for i in range(xi, xj + 1):
                for j in range(yi, yj + 1):
                    if val != grid[i][j]:
                        is_leaf = False
                        break
                if is_leaf == False:
                    break

            if not is_leaf:
                x = (xi + xj) // 2
                y = (yi + yj) // 2

                top_l = build(xi, yi, x, y)
                top_r = build(xi, y + 1, x, yj)
                bottom_l = build(x + 1, yi, xj, y)
                bottom_r = build(x + 1, y + 1, xj, yj)

                return Node(is_leaf, is_leaf, top_l, top_r, bottom_l, bottom_r)
            else:
                return Node(bool(val), is_leaf, None, None, None, None)

        n = len(grid)
        return build(0, 0, n - 1, n - 1)
